### ISSUE TITLE
Inventory of missing APIs, potential upgrade pain

### ISSUE THREAD

**Issue by unclecheese** [Firesphere/silverstripe-solr-compatibility#2]
I've taken a look at over 100 distinct implementations of `fulltextsearch` to see how inclusive this compatibility module is and where some of the sharp edges will be in practical application of upgrades from `fulltextsearch` to `silverstripe-solr-search`.

Seems like most of these could be shimmed with a PR to this module, but other APIs are open questions. 

| Deprecated Method       | Occurrences | Upgrade path | Notes                                                                                                                              |
|-------------------------|-------------|--------------|------------------------------------------------------------------------------------------------------------------------------------|
|                         |             |              |                                                                                                                                    |
| **SolrIndex**               |             |              |                                                                                                                                    |
| getTemplatesPath        | 55          | Manual       | Replaced by config. Check each implementation to make sure the value is not computed.                                              |
| getExtrasPath           | 35          | Manual       | Same as getTemplatesPath                                                                                                           |
| getTypes                | 424         | Automate     | Use custom XML. Potentially an upgrader that would render out these dynamic configs and persist them as static XML to the project. |
| addAnalyzer             | 14          | TBD          | TBD                                                                                                                                |
| getFieldDefinitions     | 600         | Automate     | Use custom XML. Potentially an upgrader that would render out these dynamic configs and persist them as static XML to the project. |
| setFieldBoosting        | 68          | Shim         | addBoostedField() is still there. This could be easily shimmed.                                                                    |
| getQueryFields          | 161         | Shim         | replaced with getFields().                                                                                                         |
| getCopyFieldDefinitions | 45          | Automate     | Use custom XML. Potentially an upgrader that would render out these dynamic configs and persist them as static XML to the project. |
| add                     | 37          | TBD          | TBD: There is not longer a public API for adding an object to an index programatically.                                            |
| delete                  | 17          | TBD          | TBD: There is not longer a public API for removing an object from an index programatically.                                        |
| commit                  | 4           | TBD          | TBD                                                                                                                                |
| search                  | 640         | Shim         | Renamed to doSearch (new params). Could be shimmed.                                                                                |
| sanitiseClassName       | 55          | Shim         | Could be shimmed                                                                                                                   |
| getFiltersComponent     | 167         | TBD          | Use solarium.                                                                                                                      |
| getService              | 670         | Shim         | Used almost entirely by _addAs                                                                                                     |
| _addAs                  | 260         | Shim         | Protected method that serves as a proxy to the Solr service.                                                                       |
| **Solr**                    |             |              |                                                                                                                                    |
| solr_options            | 27          | Shim         | Use config. Could be shimmed                                                                                                       |
| service                 | 28          | Shim         | Use injector. Could be shimmed.                                                                                                    |
|                         |             |              |                                                                                                                                    |
| **SearchQuery**             |             |              |                                                                                                                                    |
| setHandler              | 2           | Shim         | Use injector. Could be shimmed                                                                                                     |
| addFuzzySearchTerm      | 5           | Shim         | addSearchTerm exists. This could be shimmed as a wrapper around that                                                               |
| getSearchTerms          | 2           | Shim         | Simple getter. Could be shimmed                                                                                                    |
| addClassFilter          | 21          | TBD          | Class filtering isn't supported natively, but could potentially be added.                                                          |
| getExcludes             | 10          | Shim         | Renamed to getExclude(). Shimmable                                                                                                 |
| getFilters              | 10          | Shim         | Renamed to getFilter(). Shimmable                                                                                                  |
| getPageSize             | 145         | Shim         | $default_page_size exits, but it is not consumed anywhere. Appears to be an incomplete API.                                        |
| setPageSize             | 86          | Shim         | $default_page_size exits, but it is not consumed anywhere. Appears to be an incomplete API.                                        |
| addCriteria             | 3           | TBD          | Use solarium                                                                                                                       |

**Comment by Firesphere** [comment]
I won't create a project for this. These requirements rely on data that has come from FTS, and is not reliably part of the scope, nor reliably part of what a search engine should do.

**Comment by Firesphere** [comment]
Any sharp edges, should be taken care off by the CWP team and is not the responsibility of this module, nor is it the responsibility of the base module that handles Solr communication.

Upgrading anything is painful, but necessary at times. We can't just stay compliant with an old module, just for the sake of it.

**Comment by Firesphere** [comment]
**tl;dr**

I do not believe, that "upgrading from Silverstripe 3 to Silverstripe 4 is painless" is a valid statement. And therefore, also do not believe that this module, or the underlying core, should just accustom to every scenario possible. Upgrading from 3 to 4 has already proven to be a multi-week (if not multi-month) project. How is using this module such a problem, given it's documentation, making it generally about a day of work to upgrade?

**NOTE**
I am not disregarding any of these options as "possibly useful", I am however, disregarding them as part of the scope of this project.

If needed, anyone can extend this module and its requirements, to add shims as they wish. However, I do not feel it is within reason, to require any developer building anything new, to support an ancient API that is not within scope or need of the project.

# Response

After analyzing all the questions received, via Slack, Email and other ways. I have come to the conclusion that 99% of the API surface of FTS is actually used because FTS did not support something, and people had to jump through several hoops to get their search working.

This module, as it turned out, actually made it easier for them to achieve what they wanted, without copy-pasting half of the code to their own codebase.

An upgrade path, beyond the basics (e.g., method renames, if used often) is the closest I can think is our responsibility.

[The fact that the FTS module has a massive API, with a lot of pointless options, only to make sure people could actually get the results they wanted, compared to the core Solr Search module, which so far has proven to be more flexible and useful, without the need for copy paste actions, makes me believe it's not the API that needs to be implemented, but a mindset.

Upgrades are always tricky. And especially when jumping from one version to another, or from one library to another, there is no super smooth upgrade path.

I firmly believe, that the upgrade path Silverstripe suggested for upgrading form Silverstripe 3 to Silverstripe 4, was harder than this module change, if people decide this module is more useful to them.

So far, the upgrade path seems to cost about half a day (Personal experience). Which included a custom `search()` method, customised indexing because Facets were used, and custom deep fields (up to four relations deep). Yet, it took me about half a day.

Admitted, I know the inner workings of Solr and this module itself quite well, so I'm not going to claim anyone can do it within that time.

I can, however, claim that the total need for support, has taken me so far roughly a day worth of work, to get those that decided to use my core Search module completely on track. So far, about 90% of those issues were not around API changes, but around not getting things working due to being on the wrong branch, not having the latest code, or confusing FTS with the core module behind this compatibility module](https://github.com/silverstripe/silverstripe-fulltextsearch/pull/226).

# Findings

A lot of the API mentioned, have come to life due to FTS not being able to properly support Solr search. Including, but not limited to, even refusing to accept a minor change to an XML file, to support Solr 5-6 (Another set of breaking changes was introduced in Solr 7)

Yet, it is firmly believed, that devs should not be in pain for upgrading a module, yet upgrading the framework being painful is perfectly fine. I do not believe, that that is a fair outlook on the world.

Here are my responses to your list of problems:


| Deprecated Method       | Occurrences | Upgrade path | Notes                                                                                                                              | Option |
|-------------------------|-------------|--------------|------------------------------------------------------------------------------------------------------------------------------------|--------|
| **SolrIndex**               |             |              |                                                                                                                                    |  |
| getTemplatesPath        | 55          | Manual       | Replaced by config. Check each implementation to make sure the value is not computed.                                              |  |
| getExtrasPath           | 35          | Manual       | Same as getTemplatesPath                                                                                                           |  Not up to the maintainer|
| getTypes                | 424         | Automate     | Use custom XML. Potentially an upgrader that would render out these dynamic configs and persist them as static XML to the project. |Not up to the maintainer  |
| addAnalyzer             | 14          | TBD          | TBD                                                                                                                                |No reason for adding custom analyzers, they're not used in FTS anyway  |
| getFieldDefinitions     | 600         | Automate     | Use custom XML. Potentially an upgrader that would render out these dynamic configs and persist them as static XML to the project. |No reason to change  |
| setFieldBoosting        | 68          | Shim         | addBoostedField() is still there. This could be easily shimmed.                                                                    |addBoostedField is available  |
| getQueryFields          | 161         | Shim         | replaced with getFields().                                                                                                         |Use available options. There are a lot of options, none of them warrant a getQueryFields() method  |
| getCopyFieldDefinitions | 45          | Automate     | Use custom XML. Potentially an upgrader that would render out these dynamic configs and persist them as static XML to the project. |getCopyFields exists on the SchemaService. No intention of changing  |
| add                     | 37          | TBD          | TBD: There is not longer a public API for adding an object to an index programatically.                                            | add/delete/commit should not be executed by the dev |
| delete                  | 17          | TBD          | TBD: There is not longer a public API for removing an object from an index programatically.                                        | |
| commit                  | 4           | TBD          | TBD                                                                                                                                | |
| search                  | 640         | Shim         | Renamed to doSearch (new params). Could be shimmed.                                                                                | Shim exists |
| sanitiseClassName       | 55          | Shim         | Could be shimmed                                                                                                                   | Pointless |
| getFiltersComponent     | 167         | TBD          | Use solarium.                                                                                                                      | No components exists, as this module actually tries to make sense |
| getService              | 670         | Shim         | Used almost entirely by _addAs                                                                                                     | Why would this make sense in any logical way to even exist? |
| _addAs                  | 260         | Shim         | Protected method that serves as a proxy to the Solr service.                                                                       | See above |
| **Solr**                    |             |              |                                                                                                                                    | |
| solr_options            | 27          | Shim         | Use config. Could be shimmed                                                                                                       | Not needed, config is well documented and shimmed (somehow) |
| service                 | 28          | Shim         | Use injector. Could be shimmed.                                                                                                    |  Why would this even exist? |
|                         |             |              |                                                                                                                                    | |
| **SearchQuery**             |             |              |                                                                                                                                    | Why on earth would you want a custom handler? |
| setHandler              | 2           | Shim         | Use injector. Could be shimmed                                                                                                     | See below |
| addFuzzySearchTerm      | 5           | Shim         | addSearchTerm exists. This could be shimmed as a wrapper around that                                                               | Fuzzy is a setting now, which is a lot easier |
| getSearchTerms          | 2           | Shim         | Simple getter. Could be shimmed                                                                                                    | Pointless additional API surface, won't be added within this project
| addClassFilter          | 21          | TBD          | Class filtering isn't supported natively, but could potentially be added.                                                          | addFilter(ClassName::class); is available, no need for it |
| getExcludes             | 10          | Shim         | Renamed to getExclude(). Shimmable                                                                                                 | getExclude(), should not be shimmed inside this project |
| getFilters              | 10          | Shim         | Renamed to getFilter(). Shimmable                                                                                                  | getFilter(), should not be shimmed inside this project |
| getPageSize             | 145         | Shim         | $default_page_size exits, but it is not consumed anywhere. Appears to be an incomplete API.                                        | getRows() |
| setPageSize             | 86          | Shim         | $default_page_size exits, but it is not consumed anywhere. Appears to be an incomplete API.                                        | setRows() |
| addCriteria             | 3           | TBD          | Use solarium                                                                                                                       | Criteria already does this |

**Comment by Firesphere** [comment]
Closing this issue. Please make separate issues for anything that is needed within this module.

**Comment by chillu** [comment]
Wow, thanks for the in-depth analysis Simon :)

> After analyzing all the questions received, via Slack, Email and other ways. I have come to the conclusion that 99% of the API surface of FTS is actually used because FTS did not support something, and people had to jump through several hoops to get their search working.

That's a really interesting finding, I never thought about that before but it makes sense. We're still in the research phase on this, got some good customer feedback last week. As mentioned before, Solr is only one option for us in terms of what Silverstripe might support and push as a company. That' doesn't need to influence what the community can do in addition to that of course :)

**Comment by Firesphere** [comment]
Hi @chillu and @unclecheese ,

Let me clarify a few things.

The FTS module has, for example, a method to get the CopyFields. This method exists, it's just in a different location. I don't see why shimming is even needed, as copyFields can be directed from the YML, or setCopyFields or addCopyFields. The need to get them all, add some, and then put all of them back, has become pointless.

Same, for example, for the getTypes() method. This is often used to override the given types.ss template with custom types.ss. This can be set in YML, again. And I do believe that our documentation (please see the readme on the core module, it has links to the API and the documentation) is very clear about.

As for example sanitiseClassName(), it is entirely unneeded in our work, as it is sanitised by default, at every stage, and actually, it's preferred _not_ to sanitise anymore ;)

---
I know you are looking at options, and I am personally not involved in them, nor will I be. But if you choose to stay with Solr for the most of it, as I outlined before. I think it's unfair to commit to making it 100% painless, where pure framework upgrades have shown to be quite a thing.
And switching to Elastic, would probably not be "painless" either. Even if I get to 50% of the same result I've gotten with Solr so far, you can't claim that it's going to be 100% painless, in my opinion.
